cdap-app-fabric/src/main/java/io/cdap/cdap/internal/bootstrap/BootstrapService.java

```diff
@@ -102,6 +102,7 @@ protected void startUp() {
   @Override
   protected void shutDown() throws Exception {
     LOG.info("Stopping {}", getClass().getSimpleName());
+    this.systemAppManagementService.stopAndWait();
```

Comment:
it's possible this isn't started yet since it's started asynchronously. I'm not sure if it will throw an error if you try stopping a service that isn't running. Can you check the behavior?

There is also a race here where this line completes, then the thread starts the service before the executor service is shutdown. It's better to do this after the executorService is shutdown, since we know nothing will try to start it at that point.

Reply:
Documentation for stop() says "If the service has already been stopped, this method returns immediately without taking action." stopandwait() waits till service is "stopped" so my assumption is it should be no-op is service is already stopped or not started.

Also moved line according to your suggestion.

Reply:
what if it was never started? Will stopAndWait() wait forever?

Reply:
It starts out in the NEW state though, which is different than a terminal state. Can you verify that it is indeed a no-op?

Reply:
Yes for New state it is a no-op. From documentation, "If this is Service.State.NEW, it is terminated without having been started nor stopped".